File: src/mango/test/25-beginswith-test.py
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.

import copy
import mango

DOCS = [
    {"_id": "aaa", "name": "Jimi", "location": "AUS", "age": 27},
    {"_id": "abc", "name": "Eddie", "location": "AND", "age": 65},
    {"_id": "bbb", "name": "Harry", "location": "CAN", "age": 21},
    {"_id": "ccc", "name": "Eddie", "location": "DEN", "age": 37},
    {"_id": "ddd", "name": "Jones", "location": "ETH", "age": 49},
]


def to_utf8_bytes(list):
    return [x.encode() for x in list]


class BeginsWithOperator(mango.DbPerClass):
    def setUp(self):
        self.db.recreate()
        self.db.save_docs(copy.deepcopy(DOCS))
        self.db.create_index(["location"])
        self.db.create_index(["name", "location"])

    def get_mrargs(self, selector, sort=None):
        explain = self.db.find(selector, sort=sort, explain=True)
        return explain["mrargs"]

    def assertDocIds(self, user_ids, docs):
        user_ids_returned = list(d["_id"] for d in docs)
        user_ids.sort()
        user_ids_returned.sort()
        self.assertEqual(user_ids, user_ids_returned)

    def test_basic(self):
        docs = self.db.find({"location": {"$beginsWith": "A"}})

        self.assertEqual(len(docs), 2)
        self.assertDocIds(["aaa", "abc"], docs)

    def test_json_range(self):
        mrargs = self.get_mrargs({"location": {"$beginsWith": "A"}})

        self.assertEqual(mrargs["start_key"], ["A"])
        end_key_bytes = to_utf8_bytes(mrargs["end_key"])
        self.assertEqual(end_key_bytes, [b"A\xef\xbf\xbf", b"<MAX>"])

    def test_compound_key(self):
        selector = {"name": "Eddie", "location": {"$beginsWith": "A"}}
        mrargs = self.get_mrargs(selector)

        self.assertEqual(mrargs["start_key"], ["Eddie", "A"])
        end_key_bytes = to_utf8_bytes(mrargs["end_key"])
        self.assertEqual(end_key_bytes, [b"Eddie", b"A\xef\xbf\xbf", b"<MAX>"])

        docs = self.db.find(selector)
        self.assertEqual(len(docs), 1)
        self.assertDocIds(["abc"], docs)

    def test_sort(self):
        selector = {"location": {"$beginsWith": "A"}}
        cases = [
            {
                "sort": ["location"],
                "start_key": [b"A"],
                "end_key": [b"A\xef\xbf\xbf", b"<MAX>"],
                "direction": "fwd",
            },
            {
                "sort": [{"location": "desc"}],
                "start_key": [b"A\xef\xbf\xbf", b"<MAX>"],
                "end_key": [b"A"],
                "direction": "rev",
            },
        ]

        for case in cases:
            with self.subTest(sort=case["sort"]):
                mrargs = self.get_mrargs(selector, sort=case["sort"])
                self.assertEqual(to_utf8_bytes(mrargs["start_key"]), case["start_key"])
                self.assertEqual(to_utf8_bytes(mrargs["end_key"]), case["end_key"])
                self.assertEqual(mrargs["direction"], case["direction"])

    def test_all_docs_range(self):
        mrargs = self.get_mrargs({"_id": {"$beginsWith": "a"}})

        self.assertEqual(mrargs["start_key"], "a")
        end_key_bytes = to_utf8_bytes(mrargs["end_key"])
        self.assertEqual(end_key_bytes, [b"a", b"\xef\xbf\xbf"])

    def test_no_index(self):
        selector = {"foo": {"$beginsWith": "a"}}
        resp_explain = self.db.find(selector, explain=True)
        mrargs = resp_explain["mrargs"]

        self.assertEqual(resp_explain["index"]["type"], "special")
        self.assertEqual(mrargs["start_key"], None)
        self.assertEqual(mrargs["end_key"], "<MAX>")

    def test_invalid_operand(self):
        try:
            self.db.find({"_id": {"$beginsWith": True}})
        except Exception as e:
            self.assertEqual(e.response.status_code, 400)
            resp = e.response.json()
            self.assertEqual(resp["error"], "invalid_operator")
        else:
            raise AssertionError("expected find error")

    def test_does_not_match_non_string_value(self):
        docs = self.db.find({"age": {"$beginsWith": "a"}})
        self.assertEqual(len(docs), 0)

    def test_no_matches(self):
        docs = self.db.find({"name": {"$beginsWith": "Z"}})
        self.assertEqual(len(docs), 0)

    def test_case_sensitivity(self):
        docs = self.db.find({"name": {"$beginsWith": "j"}})
        self.assertEqual(len(docs), 0)

        docs = self.db.find({"name": {"$beginsWith": "J"}})
        self.assertEqual(len(docs), 2)

File: test/elixir/test/mango/25_beginswith_test.exs
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations under
# the License.

defmodule BeginsWithOperator do
  use CouchTestCase

  @db_name "begins-with-operator"

  @docs [
    %{"_id" => "aaa", "name" => "Jimi", "location" => "AUS", "age" => 27},
    %{"_id" => "abc", "name" => "Eddie", "location" => "AND", "age" => 65},
    %{"_id" => "bbb", "name" => "Harry", "location" => "CAN", "age" => 21},
    %{"_id" => "ccc", "name" => "Eddie", "location" => "DEN", "age" => 37},
    %{"_id" => "ddd", "name" => "Jones", "location" => "ETH", "age" => 49},
]

  setup do
    MangoDatabase.recreate(@db_name)
    MangoDatabase.save_docs(@db_name, @docs)
    MangoDatabase.create_index(@db_name, ["location"])
    MangoDatabase.create_index(@db_name, ["name", "location"])
    :ok
  end

  # split into Unicode graphemes
  defp split_unicode(binary) do
    binary
    |> String.graphemes()
  end

  defp get_mrargs(selector, opts \\ []) do
    sort = Keyword.get(opts, :sort, [])
    {:ok, explain} = MangoDatabase.find(@db_name, selector, sort: sort, explain: true)
    explain["mrargs"]
  end

  defp assert_doc_ids(user_ids, docs) do
    user_ids_returned =
      docs
      |> Enum.map(fn d -> d["_id"] end)
      |> Enum.sort()
    assert Enum.sort(user_ids) == user_ids_returned
  end

  test "basic" do
    {:ok, docs} = MangoDatabase.find(@db_name, %{"location" => %{"$beginsWith" => "A"}})
    assert length(docs) == 2
    assert_doc_ids(["aaa", "abc"], docs)
  end

  test "json range" do
    mrargs = get_mrargs(%{"location" => %{"$beginsWith" => "A"}})
    assert mrargs["start_key"] == ["A"]
    assert mrargs["end_key"] == [<<"A\xef\xbf\xbf">>, <<"<MAX>">>]
  end

  test "compound key" do
    selector = %{"name" => "Eddie", "location" => %{"$beginsWith" => "A"}}
    mrargs = get_mrargs(selector)
    assert mrargs["start_key"] == ["Eddie", "A"]
    assert mrargs["end_key"] == [<<"Eddie">>, <<"A\xef\xbf\xbf">>, <<"<MAX>">>]

    {:ok, docs} = MangoDatabase.find(@db_name, selector)
    assert length(docs) == 1
    assert_doc_ids(["abc"], docs)
  end

  test "sort" do
    selector = %{"location" => %{"$beginsWith" => "A"}}
    cases = [
        %{
            "sort" => ["location"],
            "start_key" => [<<"A">>],
            "end_key" => [<<"A\xef\xbf\xbf">>, <<"<MAX>">>],
            "direction" => "fwd",
        },
        %{
            "sort" => [%{"location" => "desc"}],
            "start_key" => [<<"A\xef\xbf\xbf">>, <<"<MAX>">>],
            "end_key" => [<<"A">>],
            "direction" => "rev",
        },
    ]
    for case <- cases do
      mrargs = get_mrargs(selector, sort: case["sort"])

      assert mrargs["start_key"] == case["start_key"]
      assert mrargs["end_key"] == case["end_key"]
      assert mrargs["direction"] == case["direction"]
    end
  end

  test "all docs range" do
    mrargs = get_mrargs(%{"_id" => %{"$beginsWith" => "a"}})

    assert mrargs["start_key"] == "a"
    end_key_bytes = split_unicode(mrargs["end_key"])
    assert end_key_bytes == [<<"a">>, <<"\xef\xbf\xbf">>]
  end

  test "no index" do
    selector = %{"foo" => %{"$beginsWith" => "a"}}
    {:ok, resp_explain} = MangoDatabase.find(@db_name, selector, explain: true)
    mrargs = resp_explain["mrargs"]

    assert resp_explain["index"]["type"] == "special"
    assert mrargs["start_key"] == nil
    assert mrargs["end_key"] == "<MAX>"
  end

  test "invalid operand" do
    {:error, resp} = MangoDatabase.find(@db_name, %{"_id" => %{"$beginsWith" => true}})
    assert resp.status_code == 400
    assert resp.body["error"] == "invalid_operator"
  end

  test "does not match non string value" do
    {:ok, docs} = MangoDatabase.find(@db_name, %{"age" => %{"$beginsWith" => "a"}})
    assert Enum.empty?(docs)
  end

  test "no matches" do
    {:ok, docs} = MangoDatabase.find(@db_name, %{"name" => %{"$beginsWith" => "Z"}})
    assert Enum.empty?(docs)
  end

  test "case sensitivity" do
    {:ok, docs} = MangoDatabase.find(@db_name, %{"name" => %{"$beginsWith" => "j"}})
    assert Enum.empty?(docs)

    {:ok, docs} = MangoDatabase.find(@db_name, %{"name" => %{"$beginsWith" => "J"}})
    assert length(docs) == 2
  end
end
